### PR TITLE
Move IdPSecretsProcessor class to the org.wso2.carbon.idp.mgt component

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Issue:
- https://github.com/wso2/product-is/issues/21561

With this PR, following changes will be intorduced.
- Move IdPSecretsProcessor class to the org.wso2.carbon.idp.mgt component
- Remove the org.wso2.carbon.identity.secret.mgt.core.IdPSecretsProcessor OSGI service.
- Change the IdPSecretsProcessor class usages accordingly.
- Update the unit test classes accordingly. 
- Remove application.common dependency from the secret core component.

### CODE REVIEW

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 1]
```diff
@@ -3173,13 +3176,13 @@ private IdentityProvider getIDP(Connection dbConnection, String idPName, int idp
...
                 // Retrieve encrypted secrets from DB, decrypt and set to the federated authenticator configs.
-                if (IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService() == null) {
+                if (idpSecretsProcessorService == null) {
                     throw new IdentityProviderManagementException(
                             "Error while retrieving secrets of identity provider: " + idPName + " in tenant: " +
                                     tenantDomain + ". IdPSecretsProcessorService is not available.");
                 }
                 if (federatedIdp.getFederatedAuthenticatorConfigs().length > 0) {
-                    federatedIdp = IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService().
+                    federatedIdp = idpSecretsProcessorService.
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:3185]: Can take to one line

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 2]
```diff
@@ -3901,8 +3904,8 @@ public String addIdPWithResourceId(IdentityProvider identityProvider, int tenant
 
             // Add federated authenticator secret properties to IDN_SECRET table.
             identityProvider.setId(createdIDP.getId());
-            if (IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService() != null) {
-                identityProvider = IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService().
+            if (idpSecretsProcessorService != null) {
+                identityProvider = idpSecretsProcessorService.
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:3908]: IMO null check is redundant and can take this to one line

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 3]
```diff
@@ -4244,8 +4247,8 @@ public void updateIdPWithResourceId(String resourceId, IdentityProvider
 
                 // Update secrets in IDN_SECRET table.
                 newIdentityProvider.setId(Integer.toString(idpId));
-                if (IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService() != null) {
-                    newIdentityProvider = IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService().
+                if (idpSecretsProcessorService != null) {
+                    newIdentityProvider = idpSecretsProcessorService.
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:4251]: same comment above

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 4]
```diff
@@ -4432,8 +4435,8 @@ public void deleteIdPByResourceId(String resourceId, int tenantId, String tenant
             idPName = identityProvider.getIdentityProviderName();
             deleteIdP(dbConnection, tenantId, null, resourceId);
             // Delete IdP related secrets from the IDN_SECRET table.
-            if (IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService() != null) {
-                IdpMgtServiceComponentHolder.getInstance().getIdPSecretsProcessorService().deleteAssociatedSecrets(identityProvider);
+            if (idpSecretsProcessorService != null) {
+                idpSecretsProcessorService.deleteAssociatedSecrets(identityProvider);
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:4439]: same comment above applies

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPSecretsProcessor.java [hunk 5]
```diff
@@ -16,32 +16,32 @@
...
 public class IdPSecretsProcessor implements SecretsProcessor<IdentityProvider> {
 
-    private final SecretManager secretManager;
-    private final SecretResolveManager secretResolveManager;
     private final Gson gson;
 
     public IdPSecretsProcessor() {
-
-        this.secretManager = new SecretManagerImpl();
-        this.secretResolveManager = new SecretResolveManagerImpl();
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPSecretsProcessor.java:44]: IMO, it's better to keep these references and refer the singleton instances here as each is used in multiple places over the method calls. IMO it reduce unnecessary method calls as well as improves the readability and maintainability

Comment (Thisara-Welmilla) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPSecretsProcessor.java:44]: When creating the instance of the IdPSecretsProcessor class, there can be a possibility to not to set the SecertManager and the SecretResolveManager to the `IdpMgtServiceComponentHolder` from the OSGI services.

Comment (Thisara-Welmilla) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPSecretsProcessor.java:44]: Addressed with commit b94ac1b72866b61aa03c580e3837007c501fce29

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAOTest.java [hunk 6]
```diff
@@ -136,33 +139,52 @@ private void initiateH2Database(String databaseName, String scriptPath) throws E
...
+        when(idpSecretsProcessor.decryptAssociatedSecrets(any())).thenAnswer(
+                invocation -> invocation.getArguments()[0]);
+        when(idpSecretsProcessor.encryptAssociatedSecrets(any())).thenAnswer(
+                invocation -> invocation.getArguments()[0]);
         idPManagementDAO = new IdPManagementDAO();
+        Field idpSecretsProcessorField = IdPManagementDAO.class.getDeclaredField("idpSecretsProcessorService");
+        idpSecretsProcessorField.setAccessible(true);
+        idpSecretsProcessorField.set(idPManagementDAO, idpSecretsProcessor);
+
         initiateH2Database(DB_NAME, getFilePath("h2.sql"));
```
Comment (malithie) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAOTest.java:178]: Shouldn't this be done in "BeforeClass" ?

For each test method does these need to be instantiated

Comment (Thisara-Welmilla) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAOTest.java:178]: Addressed

**Review by malithie — APPROVED** [review]

**Review by ZiyamSanthosh — APPROVED** [review]

**Pull request merged**